lib/Corres_Cases.thy

```diff
@@ -0,0 +1,323 @@
+
+  The first 3 methods take no arguments, corres_cases_both takes an optional simp argument to,
+  for example, unfold relations that synchronise cases between the abstract and concrete side.
+
+  The case distinctions work if the entire monad is a "case" statement, or if the monad is a
+  @{const bind} or @{const bindE} term with a "case" statement in the head position.
+
+  There is an existing method for case distinctions (@{method wpc}), but this method is not
+  flexible enough for @{term corres}: consider the goal
+  @{text "\<And>x. corres r (?G x) ?G' (case x of None \<Rightarrow> a | Some y \<Rightarrow> b y) m"} -- if we perform
+  case distinction on @{term x}, then we can transform @{text "?G x"} into
+  @{text "\<lambda>x s. (x = None \<longrightarrow> ?Q1 x s) \<and> (\<forall>y. x = Some y \<longrightarrow> ?Q2 x y s)"},
+  but we cannot do the same with @{text "?G'"}, because @{text "?G'"} does not depend on @{text x}.
+  The best we can do is @{text "?G' \<leadsto> \<lambda>s. ?A s \<and> ?B s"}, which so far seems to be good enough
+  in our manual proofs.
+
+  The @{method wpc} method will try to treat both preconditions uniformly and fail on @{text "?G'"}.
+  Extending @{method wpc} to deal with guards in a non-uniform way would be possible, but would make
+  setup for new constants even more messy than it already is, so we instead re-use the general idea
+  here (in Eisbach instead of ML), and leave the @{method wpc} setup clean for other uses.
+\<close>
+
+section \<open>Helper functions and definitions\<close>
+
+(* The following three definitions are originally by Dan Matichuck from the Eisbach
+   Corres_Method example *)
+
+(* Retrieve a split rule for a target term that is expected to be a case statement. *)
+ML \<open>
+fun get_split_rule ctxt target =
+let
+  val (hdTarget, args) = strip_comb (Envir.eta_contract target)
+  val (constNm, _) = dest_Const hdTarget
+  val constNm_fds = String.fields (fn c => c = #".") constNm
+
+  val _ = if String.isPrefix "case_" (List.last constNm_fds) then ()
+          else raise TERM ("Not a case statement", [target])
+
+  val typeNm = (String.concatWith "." o rev o tl o rev) constNm_fds
+  val split = Proof_Context.get_thm ctxt (typeNm ^ ".split")
+  val vars = Term.add_vars (Thm.prop_of split) []
+
+  val datatype_name = List.nth (rev constNm_fds, 1)
+
+  fun T_is_datatype (Type (nm, _)) = (Long_Name.base_name nm = Long_Name.base_name datatype_name)
+    | T_is_datatype _ = false
+
+  val datatype_var =
+    case find_first (fn ((_, _), T') => T_is_datatype T') vars of
+      SOME (ix, _) => ix
+    | NONE => error ("Couldn't find datatype in thm: " ^ datatype_name)
+
+  val split' = Drule.infer_instantiate ctxt
+                                       [(datatype_var, Thm.cterm_of ctxt (List.last args))] split
+
+in SOME split' end
+handle TERM _ => NONE;
+\<close>
+
+(* The above function as an attribute. The term argument is expected to be a case statement. *)
```

Comment:
This might be a failing of my expectations, but this comment took me quite a while to understand. My brain kept on trying to interpret 'function' as a verb, which doesn't make sense for any potential interpretation. It's probably fine to leave it as is, but I think something like `The following is the above function as an attribute` would fix it for anyone who has the same problem as me.